### DISCUSSION TITLE
Combine target dependsOn and previous --with-deps

### DISCUSSION THREAD

**Discussion by SBNTT** [nrwl/nx#10373] [General]
Hi !

Using nx v14.

Given this project.json (simplified):
```json
{
    "targets": {
        "codegen": {
            "executor": "@nrwl/workspace:run-script",
            "options": { /* ... */ }
        },
        "build": {
            "executor": "@nrwl/js:swc",
            "outputs": ["{options.outputPath}"],
            "dependsOn": ["codegen"],
            "options": { /* ... */ }
        },
    }
}
```
This module has dependency on multiple project

Without specifying the `dependsOn`, it first builds dependencies as expected.

When set the `codegen` target as a dependency of the `build` target, the actual build fails with this error:
```
Some of the project my_app's dependencies have not been built yet. Please build these libraries before:
- my_lib_a
- my_lib_b
- ...

Try: nx run my_app:build --with-deps
Some dependencies of 'my_app' have not been built. This probably due to the build target being misconfigured.
```

AFAIK, the `--with-deps` is not available anymore with nx v14

I've also tried to configure a targetDependency in the nx.json, with the exact same result.

Is there a way to specify `dependsOn` and keep nx builds dependencies automatically ?

Thanks !

**Comment by SunStupic** [comment]
I met the same issue and couldn't find related doc in nx's homepage. I found the source code in their github repo:  https://github.com/nrwl/nx/blob/4b857167cd80f73d3dd8ada0c0f5a506335f4dd0/packages/nx/src/config/nx-json.ts#L26

Try
 
```
      "dependsOn": [
        {
          "target": "build",
          "projects": "dependencies"
        },
        "codegen"
      ],
```

Acutally you can also find similar code in nx.json:

```

  "targetDependencies": {
    "build": [
      {
        "target": "build",
        "projects": "dependencies"
      }
    ]
}
```

**Reply by tbinna** [reply]
I faced a very similar issue. I think the first thing to note is that if you add the `"dependsOn": ["codegen"]` property to the `build` target, you may override the default target that is defined in the `nx.json` in your project root. In my case, the `nx.json` defined the following `targetDefaults`:

```
  "targetDefaults": {
    "build": {
      "dependsOn": ["^build"]
    }
  }
```

By overriding the default `dependsOn` configuration, your `build` target now only depends on `codegen` but not on the `build` target of dependencies anymore (`^build`). @SunStupic's suggestion fixes this by defining the build target's `dependsOn` property to `["^build", "codegen"]`.

In my case, this was not quite enough. It seems there was another issue. I was on Nx version 14.1.9 and [updated](https://nx.dev/using-nx/updating-nx) to 14.5.7 which finally did the trick.

**Reply by DaSchTour** [reply]
@tbinna thanks for this. Somehow this example is missing in the documentation to understand how this actually can be used.
